netket/jax/utils.py

```diff
@@ -320,6 +320,55 @@ def _compose(f, g):
     return reduce(_compose, funcs)
 
 
+def logsumexp(a, axis=None, b=None, keepdims=False):
+    """
+    Compute the log of the sum of exponentials of input elements.
+
+    Extended from JAX implementations to handle complex numbers.
+
+    Arguments:
+        a: (array_like) Input array.
+        axis: (None or int or tuple of ints, optional) Axis or axes over which the
+              sum is taken. By default axis is None, and all elements are summed.
+        keepdims: (bool, optional) If this is set to True, the axes which are reduced
+                  are left in the result as dimensions with size one. With this option,
+                  the result will broadcast correctly against the original array.
+        b: (array-like, optional) Scaling factor for exp(a). Must be of the same
+           shape as a or broadcastable to a.
+
+    Returns:
+        `np.log(np.sum(np.exp(a)))` calculated in a numerically more stable way.
+        If b is given then `np.log(np.sum(b*np.exp(a)))` is returned.
+    """
+    a = jnp.asarray(a)
+    result = jnp.max(a.real, axis=axis, keepdims=True)
+    if b is None:
+        result = (
+            jnp.log(
+                jnp.asarray(
+                    (jnp.exp(a - result)).sum(axis=axis, keepdims=True), dtype=complex
```

Comment:
why `dtype=complex`? I don't think you should define the dtype here.

Reply:
Also, why aren't you just using `jax.scipy.special.logsumexp`?

Reply:
> Also, why aren't you just using jax.scipy.special.logsumexp?

That version does not support complex inputs.

Is your version, @attila-i-szabo, equivalent to the scipy one wrt numerical stability? (Should we add a test maybe, to make sure?)

Reply:
Testing numerical stability is tough.
And I don't find a good implementation of `logsumexp` for complex numbers (I'm looking on the internet and not finding a reference implementation).

In any case, here we should `at_least` call into `jax.scipy` if the inputs are complex.

Reply:
> why `dtype=complex`? I don't think you should define the dtype here.

because it would break if the weight factors `b` are negative.

> Also, why aren't you just using `jax.scipy.special.logsumexp`?

because it doesn't work properly with complex numbers. That's like 99% of the functionality of `nkjax`...

> Is your version, @attila-i-szabo, equivalent to the scipy one wrt numerical stability? (Should we add a test maybe, to make sure?)

Apart from the dtype specification (and some stuff they need to ensure the output is real), the implementation is identical to that one

Reply:
I'm still not sure about this. 
I'm seriously not sure this is the best way to make it work.
`Jax.scipy.special.logsumexp` requires a one line change (`lax.is_finite`->`jnp.isfinite`) to work with complex numbers. 
I'll try to open a PR to jax about it. 
In the meantime I'd prefer this change (and related stuff) not to be in here and to be in another PR to be discussed.

Reply:
For reference, https://github.com/google/jax/pull/6855

Reply:
OK, it stays here for now because other stuff relies on it, and it's useful for NQS development in general (like much of the rest of these utils). If your PR gets added to JAX, we can scrap it by changing about two lines.

Reply:
No. 
I don't agree with this implementation because it promotes everything to complex numbers.
It would not be compatible with my PR to jax.
It does something different from what a standard implementation of logsumexp does and thus should be named differently to avoid confusion.

This change does not fit in this PR (which is already huge). 
Please remove it and let's discuss it in another PR.

Reply:
My PR was accepted so it will most probably end up in the next jax release 2 weeks from now.
I think the best solution would be to copy-paste the code of `jax.scipy.special.logsumexp` somewhere in netket and use that unless there is a (Future) version of netket.

But this and the changes in `nn/linear.py` in another PR.